In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
import numpy as np 
from sklearn import linear_model
import seaborn as seabornInstance
from sklearn.linear_model import LinearRegression
from sklearn import metrics
import plotly.graph_objects as go
%matplotlib inline


In [2]:
bowling_m=pd.read_excel('bowling_ml_letsgo.xlsx')

In [3]:
bowling_m.head()

Unnamed: 0  height  exp  pro_exp       birth  weight  ball_pound  \
0           0     180   23        3  1986-07-03      60        15.0   
1           1     173   22        3  1984-10-03      79        15.0   
2           2     179   13        4  1993-10-24     130        15.0   
3           3     175   10        8  1981-08-01      72        15.0   
4           5     174   26       10  1976-09-28      74        15.0   

   left_hand  right_hand  cranker  stroker  twohanded  thumbless    CON_ST  \
0          1           0        1        0          0          0  4.400000   
1          0           1        1        0          0          0  5.750000   
2          0           1        1        0          0          0  2.750000   
3          0           1        1        0          0          0  4.000000   
4          0           1        1        0          0          0  3.333333   

     L_READ  OVERCOME  SPARE_SUC     TOT_AVG  
0 -9.000000  0.652252   0.811789  215.800000  
1 -9.750000  0.687281   0.659609  246.000000  
2 -9.250000  0.475010   0.542429  190.500000  
3 -8.750000  0.455561   0.768218  212.500000  
4 -8.666667  0.783212   0.624562  198.333333

In [4]:
bowling_m_2=bowling_m.drop(['Unnamed: 0','birth'],axis=1)

In [6]:
bowling_m_2.head()

height  exp  pro_exp  weight  ball_pound  left_hand  right_hand  cranker  \
0     180   23        3      60        15.0          1           0        1   
1     173   22        3      79        15.0          0           1        1   
2     179   13        4     130        15.0          0           1        1   
3     175   10        8      72        15.0          0           1        1   
4     174   26       10      74        15.0          0           1        1   

   stroker  twohanded  thumbless    CON_ST    L_READ  OVERCOME  SPARE_SUC  \
0        0          0          0  4.400000 -9.000000  0.652252   0.811789   
1        0          0          0  5.750000 -9.750000  0.687281   0.659609   
2        0          0          0  2.750000 -9.250000  0.475010   0.542429   
3        0          0          0  4.000000 -8.750000  0.455561   0.768218   
4        0          0          0  3.333333 -8.666667  0.783212   0.624562   

      TOT_AVG  
0  215.800000  
1  246.000000  
2  190.500000  
3  212.500000  
4  198.333333

---
## ☛ 1. columns의 모든 피처를 사용해도 괜찮을까?
---

In [5]:
X=bowling_m_2.drop(['TOT_AVG'],axis=1)
y=bowling_m_2['TOT_AVG']

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)

In [7]:
lm= sm.OLS(y_train, X_train).fit()
lm.summary()

c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\scipy\stats\stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                TOT_AVG   R-squared:                       0.911
Model:                            OLS   Adj. R-squared:                  0.689
Method:                 Least Squares   F-statistic:                     4.098
Date:                Wed, 11 May 2022   Prob (F-statistic):             0.0932
Time:                        19:04:20   Log-Likelihood:                -42.685
No. Observations:                  15   AIC:                             107.4
Df Residuals:                       4   BIC:                             115.2
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
height        -0.9109      0.708     -1.287      0.268      -2.877       1.055
exp           -0.5386      0.399     -1.350      0.248      -1.647       0.569
pro_exp       -0.1518      0.580     -0.262      0.806      -1.761       1.457
weight        -0.1938      0.171     -1.130      0.322      -0.670       0.282
ball_pound    13.7175     13.825      0.992      0.377     -24.667      52.102
left_hand     35.7400     77.926      0.459      0.670    -180.616     252.096
right_hand    34.6696     80.230      0.432      0.688    -188.086     257.425
cranker       70.4096    158.023      0.446      0.679    -368.333     509.153
stroker     3.847e-14   9.69e-14      0.397      0.712   -2.31e-13    3.08e-13
twohanded  -8.539e-16   1.44e-15     -0.592      0.586   -4.86e-15    3.15e-15
thumbless           0          0        nan        nan           0           0
CON_ST        11.1882      5.129      2.181      0.095      -3.053      25.430
L_READ        -2.8043      4.218     -0.665      0.543     -14.515       8.907
OVERCOME      22.4597     14.361      1.564      0.193     -17.413      62.332
SPARE_SUC      7.9967     24.804      0.322      0.763     -60.869      76.863
==============================================================================
Omnibus:                        0.655   Durbin-Watson:                   2.444
Prob(Omnibus):                  0.721   Jarque-Bera (JB):                0.675
Skew:                          -0.327   Prob(JB):                        0.714
Kurtosis:                       2.192   Cond. No.                     3.06e+20
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.21e-36. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [8]:
# 4개의 생성 피처만 넣으면?

X=bowling_m_2[['CON_ST','L_READ','OVERCOME','SPARE_SUC']]
y=bowling_m_2['TOT_AVG']

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)
lm= sm.OLS(y_train, X_train).fit()
lm.summary()

c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\scipy\stats\stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                TOT_AVG   R-squared (uncentered):                   0.998
Model:                            OLS   Adj. R-squared (uncentered):              0.997
Method:                 Least Squares   F-statistic:                              1476.
Date:                Wed, 11 May 2022   Prob (F-statistic):                    6.22e-15
Time:                        19:04:30   Log-Likelihood:                         -54.768
No. Observations:                  15   AIC:                                      117.5
Df Residuals:                      11   BIC:                                      120.4
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CON_ST        10.9726      3.689      2.974      0.013       2.853      19.092
L_READ       -13.0948      2.285     -5.731      0.000     -18.124      -8.066
OVERCOME      37.1912     15.789      2.355      0.038       2.439      71.943
SPARE_SUC     41.7445     23.022      1.813      0.097      -8.926      92.415
==============================================================================
Omnibus:                        2.409   Durbin-Watson:                   2.712
Prob(Omnibus):                  0.300   Jarque-Bera (JB):                1.283
Skew:                           0.716   Prob(JB):                        0.527
Kurtosis:                       2.972   Cond. No.                         82.7
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [10]:
# 4개외의 피처만 넣으면?

X=bowling_m_2.drop(['CON_ST','L_READ','OVERCOME','SPARE_SUC','TOT_AVG'],axis=1)
y=bowling_m_2['TOT_AVG']

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)
lm= sm.OLS(y_train, X_train).fit()
lm.summary()

c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\scipy\stats\stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "
c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\statsmodels\regression\linear_model.py:1918: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                TOT_AVG   R-squared:                       0.716
Model:                            OLS   Adj. R-squared:                  0.503
Method:                 Least Squares   F-statistic:                     3.365
Date:                Wed, 11 May 2022   Prob (F-statistic):             0.0584
Time:                        19:04:41   Log-Likelihood:                -51.389
No. Observations:                  15   AIC:                             116.8
Df Residuals:                       8   BIC:                             121.7
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
height        -0.4949      0.874     -0.566      0.587      -2.510       1.520
exp           -0.3364      0.447     -0.753      0.473      -1.366       0.694
pro_exp       -0.3060      0.691     -0.443      0.670      -1.899       1.287
weight        -0.3878      0.177     -2.195      0.059      -0.795       0.020
ball_pound    34.4853     10.833      3.183      0.013       9.503      59.467
left_hand    -55.2378     80.057     -0.690      0.510    -239.848     129.373
right_hand   -63.5086     81.641     -0.778      0.459    -251.773     124.756
cranker     -118.7464    161.576     -0.735      0.483    -491.340     253.848
stroker             0          0        nan        nan           0           0
twohanded           0          0        nan        nan           0           0
thumbless           0          0        nan        nan           0           0
==============================================================================
Omnibus:                        1.245   Durbin-Watson:                   2.626
Prob(Omnibus):                  0.537   Jarque-Bera (JB):                0.922
Skew:                           0.333   Prob(JB):                        0.631
Kurtosis:                       1.984   Cond. No.                          inf
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [11]:
#4개 피처+볼파운드 는?

X=bowling_m_2[['CON_ST','L_READ','OVERCOME','SPARE_SUC','ball_pound']]
y=bowling_m_2['TOT_AVG']


X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)
lm= sm.OLS(y_train, X_train).fit()
lm.summary()

c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\scipy\stats\stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                TOT_AVG   R-squared (uncentered):                   0.999
Model:                            OLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                              2042.
Date:                Wed, 11 May 2022   Prob (F-statistic):                    1.05e-14
Time:                        19:05:22   Log-Likelihood:                         -49.954
No. Observations:                  15   AIC:                                      109.9
Df Residuals:                      10   BIC:                                      113.4
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CON_ST         8.8888      2.892      3.074      0.012       2.446      15.332
L_READ        -4.0375      3.484     -1.159      0.273     -11.800       3.725
OVERCOME      24.8931     12.693      1.961      0.078      -3.390      53.176
SPARE_SUC     25.1769     18.366      1.371      0.200     -15.745      66.099
ball_pound     7.2455      2.415      3.000      0.013       1.865      12.626
==============================================================================
Omnibus:                        1.343   Durbin-Watson:                   2.076
Prob(Omnibus):                  0.511   Jarque-Bera (JB):                0.918
Skew:                           0.292   Prob(JB):                        0.632
Kurtosis:                       1.937   Cond. No.                         159.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [58]:
#CON_ST, ball_pound 사용시엔 ?


X=bowling_m_2[['CON_ST','ball_pound']]
y=bowling_m_2['TOT_AVG']


X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)
lm= sm.OLS(y_train, X_train).fit()
lm.summary()

c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\scipy\stats\stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                TOT_AVG   R-squared (uncentered):                   0.998
Model:                            OLS   Adj. R-squared (uncentered):              0.998
Method:                 Least Squares   F-statistic:                              4143.
Date:                Wed, 11 May 2022   Prob (F-statistic):                    5.85e-19
Time:                        14:39:24   Log-Likelihood:                         -53.482
No. Observations:                  15   AIC:                                      111.0
Df Residuals:                      13   BIC:                                      112.4
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CON_ST         9.8812      3.165      3.122      0.008       3.044      16.719
ball_pound    11.5237      0.870     13.240      0.000       9.643      13.404
==============================================================================
Omnibus:                        0.161   Durbin-Watson:                   2.128
Prob(Omnibus):                  0.923   Jarque-Bera (JB):                0.042
Skew:                           0.036   Prob(JB):                        0.979
Kurtosis:                       2.751   Cond. No.                         21.8
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

--- 
### 📍 정리

 1. 모든 피처 -> 4개의 피처
       - R-squared(Adj. R-squared) 가 모두 크고 1에 더 가까움
       - Prob (F-statistic) 가 0.05 이하 매우 작음  
       
              - but :  AIC ,BIC 값이 조금 더 상승함...   
                

                
2. 4개 제외 피처(선수관련 특성)  

       - R-squared(Adj. R-squared) 낮아지고 Prob(F-statistic) 가 높아짐  
    
             - but : 특성 중 ball_pound 유의미 한 특성  
              
    
3. 4개+ ball_pound  

       - R-squared(Adj. R-squared) 가 모두 크고 1에 더 가까움  

       - Prob (F-statistic) 가 0.05 이하 매우 작음  
       
       - 1. AIC ,BIC 값 낮아짐  
       



  > 결론 :    피처  ['CON_ST' , 'L_READ' , 'OVERCOME' , 'SPARE_SUC' , 'ball_pound']  사용하자

  ---

  
  
---
## ☛ 2. 다중선형회귀분석
---

In [46]:
X=bowling_m_2[['CON_ST','L_READ','OVERCOME','SPARE_SUC','ball_pound']]
y=bowling_m_2['TOT_AVG']
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)

In [13]:
X_train.shape,y_train.shape, X_test.shape, y_test.shape


((15, 5), (15,), (4, 5), (4,))

In [47]:
#회귀분석 모델 학습

lr= linear_model.LinearRegression()
lr_model=lr.fit(X_train, y_train)

In [48]:
#회귀 분석 모델 평가(R2 score :1에 가까울 수록)

print('학습 데이터셋의 R2 score : ',lr_model.score(X_train, y_train))
print('테스트 데이터셋의 R2 score : ',lr_model.score(X_test, y_test))

학습 데이터셋의 R2 score :  0.7791523870170818
테스트 데이터셋의 R2 score :  0.9389558085279618


In [49]:
#회귀 분석 모델 평가(RMSE score: 예측값과 실제값과의 차이, 작을수록)

y_pred=lr.predict(X_train)
print('학습 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_train, y_pred)))
y_pred=lr.predict(X_test)
print('테스트 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_test, y_pred)))

학습 데이터셋의 RMSE :  6.563805935325276
테스트 데이터셋의 RMSE :  4.682584966445904


--- 
### 📍 정리

1. 학습 데이터와 테스트 데이터의 평가가 다소 차이 남  

    - 8:2 > 7:3으로 수정한다면? 어떻게 바뀔것인가  
    -
    
    
    
2. RMSE를 더 낮추고 싶음  

    - 피처스케일링을 적용해볼까?  
    -
                


  > 결론 :  
      1) train/test 크기 변경 후 확인  
      2) 피쳐 스케일링 후 다시 회귀

  ---

  
  
---
## ☛ 3. 수정 후 다시 분석
---

In [50]:
# 1. 학습/테스트 7:3

X=bowling_m_2[['CON_ST','L_READ','OVERCOME','SPARE_SUC','ball_pound']]
y=bowling_m_2['TOT_AVG']
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3, random_state=13)


lr= linear_model.LinearRegression()
lr_model=lr.fit(X_train, y_train)

In [51]:
from sklearn.metrics import mean_squared_error
from math import sqrt



print('학습 데이터셋의 R2 score : ',lr_model.score(X_train, y_train))
print('테스트 데이터셋의 R2 score : ',lr_model.score(X_test, y_test))

y_pred=lr.predict(X_train)
print('학습 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_train, y_pred)))
y_pred=lr.predict(X_test)
print('테스트 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_test, y_pred)))

학습 데이터셋의 R2 score :  0.7995025077331275
테스트 데이터셋의 R2 score :  0.8143175094604066
학습 데이터셋의 RMSE :  6.151834759202189
테스트 데이터셋의 RMSE :  7.853187581221425


In [ ]:
#해석
학습/테스트 랜덤분류 > 모델에 영향x
학습-테스트 데이터의 크기 차이가 줄어들 수록  score 차가 작아질 뿐.
유의미한 결론x

In [21]:
bowling_m_3=bowling_m_2[['CON_ST','L_READ','OVERCOME','SPARE_SUC','ball_pound','TOT_AVG']]
bowling_m_3.head()

CON_ST    L_READ  OVERCOME  SPARE_SUC  ball_pound     TOT_AVG
0  4.400000 -9.000000  0.652252   0.811789        15.0  215.800000
1  5.750000 -9.750000  0.687281   0.659609        15.0  246.000000
2  2.750000 -9.250000  0.475010   0.542429        15.0  190.500000
3  4.000000 -8.750000  0.455561   0.768218        15.0  212.500000
4  3.333333 -8.666667  0.783212   0.624562        15.0  198.333333

In [24]:
# 2. 피쳐 스케일링 - boxplot 확인
fig= go.Figure()
fig.add_traces(go.Box(y=bowling_m_3['CON_ST'],name='CON_ST'))
fig.add_traces(go.Box(y=bowling_m_3['L_READ'],name='L_READ'))
fig.add_traces(go.Box(y=bowling_m_3['OVERCOME'],name='OVERCOME'))
fig.add_traces(go.Box(y=bowling_m_3['SPARE_SUC'],name='SPARE_SUC'))
fig.add_traces(go.Box(y=bowling_m_3['ball_pound'],name='ball_pound'))
fig.add_traces(go.Box(y=bowling_m_3['TOT_AVG'],name='TOT_AVG'))

fig.show()

In [25]:
fig= go.Figure()
fig.add_traces(go.Box(y=bowling_m_3['CON_ST'],name='CON_ST'))
fig.add_traces(go.Box(y=bowling_m_3['L_READ'],name='L_READ'))
fig.add_traces(go.Box(y=bowling_m_3['OVERCOME'],name='OVERCOME'))
fig.add_traces(go.Box(y=bowling_m_3['SPARE_SUC'],name='SPARE_SUC'))
fig.add_traces(go.Box(y=bowling_m_3['ball_pound'],name='ball_pound'))


fig.show()

In [28]:
# 2. 피쳐 스케일링
from sklearn.preprocessing import MinMaxScaler, StandardScaler


mms=MinMaxScaler()
ss=StandardScaler()

bowling_mms=mms.fit_transform(bowling_m_3)
bowling_mms_pd = pd.DataFrame(bowling_mms, columns=bowling_m_3.columns)


bowling_ss=ss.fit_transform(bowling_m_3)
bowling_ss_pd = pd.DataFrame(bowling_ss, columns=bowling_m_3.columns)


In [29]:
bowling_mms_pd.head()

CON_ST    L_READ  OVERCOME  SPARE_SUC  ball_pound   TOT_AVG
0  0.507692  0.571429  0.548268   0.973629         0.0  0.430028
1  0.923077  0.250000  0.604706   0.568044         0.0  0.943343
2  0.000000  0.464286  0.262702   0.255741         0.0  0.000000
3  0.384615  0.678571  0.231366   0.857506         0.0  0.373938
4  0.179487  0.714286  0.759266   0.474638         0.0  0.133144

In [56]:
fig= go.Figure()
fig.add_traces(go.Box(y=bowling_mms_pd['CON_ST'],name='CON_ST'))
fig.add_traces(go.Box(y=bowling_mms_pd['L_READ'],name='L_READ'))
fig.add_traces(go.Box(y=bowling_mms_pd['OVERCOME'],name='OVERCOME'))
fig.add_traces(go.Box(y=bowling_mms_pd['SPARE_SUC'],name='SPARE_SUC'))
fig.add_traces(go.Box(y=bowling_mms_pd['ball_pound'],name='ball_pound'))
fig.add_traces(go.Box(y=bowling_mms_pd['TOT_AVG'],name='TOT_AVG'))


In [54]:
bowling_ss_pd.head()

CON_ST    L_READ  OVERCOME  SPARE_SUC  ball_pound   TOT_AVG
0  0.234471  0.082023  0.376106   1.175288   -0.433013 -0.066266
1  1.725594 -0.996890  0.585194  -0.172963   -0.433013  1.907269
2 -1.588011 -0.277615 -0.681827  -1.211123   -0.433013 -1.719592
3 -0.207343  0.441660 -0.797915   0.789270   -0.433013 -0.281917
4 -0.943699  0.561539  1.157788  -0.483462   -0.433013 -1.207692

In [57]:
fig= go.Figure()
fig.add_traces(go.Box(y=bowling_ss_pd['CON_ST'],name='CON_ST'))
fig.add_traces(go.Box(y=bowling_ss_pd['L_READ'],name='L_READ'))
fig.add_traces(go.Box(y=bowling_ss_pd['OVERCOME'],name='OVERCOME'))
fig.add_traces(go.Box(y=bowling_ss_pd['SPARE_SUC'],name='SPARE_SUC'))
fig.add_traces(go.Box(y=bowling_ss_pd['ball_pound'],name='ball_pound'))
fig.add_traces(go.Box(y=bowling_ss_pd['TOT_AVG'],name='TOT_AVG'))

In [58]:
# 2-1) min-max scaler >> 다중선형회귀


X=bowling_mms_pd.drop(['TOT_AVG'],axis=1)
y=bowling_mms_pd['TOT_AVG']
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)


lr= linear_model.LinearRegression()
lr_model=lr.fit(X_train, y_train)


In [59]:
#에러 계산 (동일)

# pred_tr=lr.predict(X_train)
# pred_test=lr.predict(X_test)

# rmse_tr = (np.sqrt(mean_squared_error(y_train, pred_tr)))
# rmse_test = (np.sqrt(mean_squared_error(y_test, pred_test)))

# print("RMSE of train data: ",rmse_tr)
# print("RMSE of test data: ",rmse_test)

RMSE of train data:  0.11156610654943816
RMSE of test data:  0.07959067931636159


In [60]:

print('학습 데이터셋의 R2 score : ',lr_model.score(X_train, y_train))
print('테스트 데이터셋의 R2 score : ',lr_model.score(X_test, y_test))

y_pred=lr.predict(X_train)
print('학습 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_train, y_pred)))
y_pred=lr.predict(X_test)
print('테스트 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_test, y_pred)))

학습 데이터셋의 R2 score :  0.779152387017082
테스트 데이터셋의 R2 score :  0.938955808527961
학습 데이터셋의 RMSE :  0.11156610654943816
테스트 데이터셋의 RMSE :  0.07959067931636159


In [ ]:
#scaling 이전에 비해  R2 score 동일 / RMSE가 감소 

In [61]:
# 2-2) standard scaler >> 다중선형회귀

X=bowling_ss_pd.drop(['TOT_AVG'],axis=1)
y=bowling_ss_pd['TOT_AVG']
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)


lr= linear_model.LinearRegression()
lr_model=lr.fit(X_train, y_train)


In [62]:

print('학습 데이터셋의 R2 score : ',lr_model.score(X_train, y_train))
print('테스트 데이터셋의 R2 score : ',lr_model.score(X_test, y_test))

y_pred=lr.predict(X_train)
print('학습 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_train, y_pred)))
y_pred=lr.predict(X_test)
print('테스트 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_test, y_pred)))




학습 데이터셋의 R2 score :  0.779152387017082
테스트 데이터셋의 R2 score :  0.9389558085279612
학습 데이터셋의 RMSE :  0.42893711067145923
테스트 데이터셋의 RMSE :  0.30600150061892284


In [ ]:
#scaling 이전에 비해  R2 score 동일 / RMSE가 감소 
# mms에 비해 RMSE가 증가

>>그럼 MinMaxScaler 사용

In [63]:
X=bowling_mms_pd.drop(['TOT_AVG'],axis=1)
y=bowling_mms_pd['TOT_AVG']
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)

lm= sm.OLS(y_train, X_train).fit()
lm.summary()

c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\scipy\stats\stats.py:1541: UserWarning:

kurtosistest only valid for n>=20 ... continuing anyway, n=15



<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                TOT_AVG   R-squared (uncentered):                   0.947
Model:                            OLS   Adj. R-squared (uncentered):              0.920
Method:                 Least Squares   F-statistic:                              35.62
Date:                Wed, 11 May 2022   Prob (F-statistic):                    4.65e-06
Time:                        20:32:52   Log-Likelihood:                          11.427
No. Observations:                  15   AIC:                                     -12.85
Df Residuals:                      10   BIC:                                     -9.313
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CON_ST         0.6370      0.226      2.820      0.018       0.134       1.140
L_READ        -0.1441      0.115     -1.253      0.239      -0.400       0.112
OVERCOME       0.2565      0.126      2.039      0.069      -0.024       0.537
SPARE_SUC      0.1728      0.099      1.752      0.110      -0.047       0.393
ball_pound     0.0176      0.153      0.115      0.911      -0.324       0.359
==============================================================================
Omnibus:                        1.932   Durbin-Watson:                   2.025
Prob(Omnibus):                  0.381   Jarque-Bera (JB):                0.950
Skew:                           0.110   Prob(JB):                        0.622
Kurtosis:                       1.787   Cond. No.                         8.33
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
# 피처 스케일링 이전에 비해 
# Adj. R-squared 낮아짐
# AIC,BIC 음수?
# p-value 변화 +-

In [64]:
X=bowling_m_2[['CON_ST','L_READ','OVERCOME','SPARE_SUC','ball_pound']]
y=bowling_m_2['TOT_AVG']


X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)
lm= sm.OLS(y_train, X_train).fit()
lm.summary()

c:\Users\user\Anaconda3\envs\ds_study\lib\site-packages\scipy\stats\stats.py:1541: UserWarning:

kurtosistest only valid for n>=20 ... continuing anyway, n=15



<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                TOT_AVG   R-squared (uncentered):                   0.999
Model:                            OLS   Adj. R-squared (uncentered):              0.999
Method:                 Least Squares   F-statistic:                              2042.
Date:                Wed, 11 May 2022   Prob (F-statistic):                    1.05e-14
Time:                        20:33:48   Log-Likelihood:                         -49.954
No. Observations:                  15   AIC:                                      109.9
Df Residuals:                      10   BIC:                                      113.4
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CON_ST         8.8888      2.892      3.074      0.012       2.446      15.332
L_READ        -4.0375      3.484     -1.159      0.273     -11.800       3.725
OVERCOME      24.8931     12.693      1.961      0.078      -3.390      53.176
SPARE_SUC     25.1769     18.366      1.371      0.200     -15.745      66.099
ball_pound     7.2455      2.415      3.000      0.013       1.865      12.626
==============================================================================
Omnibus:                        1.343   Durbin-Watson:                   2.076
Prob(Omnibus):                  0.511   Jarque-Bera (JB):                0.918
Skew:                           0.292   Prob(JB):                        0.632
Kurtosis:                       1.937   Cond. No.                         159.
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [72]:
#그럼 단순선형회귀는? 크게 차이 날까? CON_ST, SPARE_SUC

# 1)CON_ST

X=bowling_m_2[['CON_ST']]
y=bowling_m_2['TOT_AVG']
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)

In [73]:
lr= linear_model.LinearRegression()
lr_model=lr.fit(X_train, y_train)

In [69]:
print('학습 데이터셋의 R2 score : ',lr_model.score(X_train, y_train))
print('테스트 데이터셋의 R2 score : ',lr_model.score(X_test, y_test))
y_pred=lr.predict(X_train)
print('학습 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_train, y_pred)))
y_pred=lr.predict(X_test)
print('테스트 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_test, y_pred)))

학습 데이터셋의 R2 score :  0.660887499490185
테스트 데이터셋의 R2 score :  0.9342983031458743
학습 데이터셋의 RMSE :  8.133571281088562
테스트 데이터셋의 RMSE :  4.857935973102937


In [74]:
# 2) SPARE_SUC
X=bowling_m_2[['SPARE_SUC']]
y=bowling_m_2['TOT_AVG']
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=13)

In [75]:
lr= linear_model.LinearRegression()
lr_model=lr.fit(X_train, y_train)

In [76]:
print('학습 데이터셋의 R2 score : ',lr_model.score(X_train, y_train))
print('테스트 데이터셋의 R2 score : ',lr_model.score(X_test, y_test))
y_pred=lr.predict(X_train)
print('학습 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_train, y_pred)))
y_pred=lr.predict(X_test)
print('테스트 데이터셋의 RMSE : ', sqrt(mean_squared_error(y_test, y_pred)))

학습 데이터셋의 R2 score :  0.10635359237743802
테스트 데이터셋의 R2 score :  -0.26770798944649954
학습 데이터셋의 RMSE :  13.20359621243556
테스트 데이터셋의 RMSE :  21.33895301753605


In [ ]:
결론 : 앗! 여러 피처선택이 맞음!


## 결론

**1. 5개 피처 사용 [['CON_ST','L_READ','OVERCOME','SPARE_SUC','ball_pound']]**  

	 - 학습/테스트 R2 :  0.7791523870170818 / 0.9389558085279618  
	 - 학습/테스트 RMSE :  6.563805935325276 / 4.682584966445904  



**2. 5개 피처 + MinMax scaler**  

	 - 학습/테스트 R2 :  0.7791523870170818 / 0.9389558085279618  
	 - 학습/테스트 RMSE : 0.11156610654943816 / 0.07959067931636159  

	

**3. 5개 피처 + Standard scaler**  

	 - 학습/테스트 R2 :  0.7791523870170818 / 0.9389558085279618  
	 - 학습/테스트 RMSE :0.42893711067145923 / 0.30600150061892284